Cad1task1 baseline2

groadabike opened the pull request:
This PR for release v0.3.2

What's included?

In Cad 1 - Task 1:
* Second baseline 1 using the spectrogram-based Open-UnmIx model.  
* Saves VDBO stems at 24 kHz sample rate.
* Saves Remix signals at 32 kHz sample rate.
* Saves all signals using 16bit.
* Compresses and saves all signals using FLAC (https://en.wikipedia.org/wiki/FLAC).

In Cad1 - Task 2:
* Saves signals at 32 kHz sample rate.
* Saves signals using 16bit.
* Compresses and saves all signals using FLAC

Other changes:

jonbarker68 commented on recipes/cad1/task1/baseline/enhance.py:
```diff
@@ -142,10 +143,12 @@ def map_to_dict(sources: np.ndarray, sources_list: list[str]) -> dict:
 
 # pylint: disable=unused-argument
 def decompose_signal(
```
I'd suggest moving the model_name logic to the calling function. Having both model and model_name passed to this lower level function invite misuse, i.e. passing a demucs model but setting name to be something else.  This function is basically: do some optional normalization, do some optional resampling and then separate the signals. It shouldn't have the model name as a parameter.  The enhance() function should be using this name to load the correct model and then to set up the call to this function.

groadabike replied:
Yeah, I was unsure about how to deal with the model name. 
I may need to move the resampling and normalization to enhance()

jonbarker68 commented on recipes/cad1/task1/baseline/enhance.py:
```diff
@@ -247,6 +265,47 @@ def process_stems_for_listener(
 
 
+def clip_signal(signal: np.ndarray, soft_clip: bool = False) -> tuple[np.ndarray, int]:
+    """Clip and save the processed stems.
```
Comment says "clip and save" but the function is returning the clipped signal

jonbarker68 commented on recipes/cad1/task1/baseline/enhance.py:
```diff
@@ -395,15 +469,11 @@ def enhance(config: DictConfig) -> None:
+        clipped_signal, n_clipped = clip_signal(enhanced, config.soft_clip)
         if n_clipped > 0:
             logger.warning(f"Writing {filename}: {n_clipped} samples clipped")
-        np.clip(enhanced, -1.0, 1.0, out=enhanced)
```
This is duplicated code. We have a write_signal that handles clipping and conversion to 16bit.  The only bit you might need to do would be the

Something like

```
if config.soft_clip:
  signal = tanh(signal)
write_signal(signal, filename)
```

You can then remove the clip_signal, to_16bit and resample functions to avoid duplicating code.

The read_signal write_signal  methods were recently refactored in the main branch. So if the plan is to release this as 0.3.2 it would be worth copying them over from main now.

I think it's important that we try to reuse our library code in the recipes wherever possible so that others can see by example.

ns-rse replied:
Agree with reusing code. I think it should be possible to `git cherry-pick` the merge commit from `main` that incorporated these changes on that branch to pull them all over.

groadabike replied:
Ok, I didn't reuse these functions as they are quite "new" in main.
I wasn't sure how to deal with all these defferences between 0.3 and main

groadabike replied:
Ok, I didn't reuse these functions as they are quite "new" in main.
I wasn't sure how to deal with all these defferences between 0.3 and main

groadabike replied:
@ns-rse, I always confused about the cherry-pick.
What would be better in this case, `charry-pick` or just add another new commit copying these function?

groadabike replied:
@ns-rse, I always confused about the cherry-pick.
What would be better in this case, `cherry-pick` or just add another new commit copying these function?

groadabike replied:
@ns-rse, I always confused about the cherry-pick.
What would be better in this case, `charry-pick` or just add another new commit copying these function?

ns-rse replied:
I don't have a particularly strong opinion, I think it depends whether you want to have _everything_ tracked in the `git` history, which `cherry-pick` should preserve.

If its straight-forward to copy code manually from another branch and commit and you're comfortable doing that it would also be fine.

jonbarker68 reviewed (commented):
Thanks for adding this. Will be great to have a couple of different approaches. I've added a few suggestions.

jonbarker68 commented:
I guess there is a danger of complicating things here.  We probably want to
keep changes on 0.3.x as minimal as possible and constrained to the recipe
code.

I would keep the changes to the recipe in 0.3.2 as small as possible, i.e.
stick to adding the new demixing functionality. Then we'll refactor all the
recipes to make better use of the existing write_signal code in the ongoing
refactoring work that will be released in August for  0.4.0

(i.e. the write_signal code duplication is already in the 0.3. release so
this small update is not the place to be fixing it)

Jon

On Thu, 20 Apr 2023 at 10:55, Gerardo Roa Dabike ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In recipes/cad1/task1/baseline/enhance.py
> <https://github.com/claritychallenge/clarity/pull/300#discussion_r1172361182>
> :
>
> >          if n_clipped > 0:
>              logger.warning(f"Writing {filename}: {n_clipped} samples clipped")
> -        np.clip(enhanced, -1.0, 1.0, out=enhanced)
>
> @ns-rse <https://github.com/ns-rse>, I always confused about the
> cherry-pick.
> What would be better in this case, charry-pick or just add another new
> commit copying these function?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/claritychallenge/clarity/pull/300#discussion_r1172361182>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO7SJABAO4CKBGQ7NNYO4TXCEBZPANCNFSM6AAAAAAXELDSJY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Professor Jon Barker,
Department of Computer Science,
University of Sheffield
+44 (0) 114 222 1824

jonbarker68 commented on recipes/cad1/task1/baseline/enhance.py:
```diff
@@ -247,6 +250,47 @@ def process_stems_for_listener(
+
+
 @hydra.main(config_path="", config_name="config")
 def enhance(config: DictConfig) -> None:
```
To make it easier to extend, we can put all the technique-dependent logic in one place, e.g. like below, and place the normalise/denormalise logic into decompose_signal function,  which will have an extra boolean parameter 'normalise'.  So the setup logic will look like

```
            if config.separator.model == "demucs":
                model_sample_rate = 44100
                sources_order = separation_model.sources
                normalise = True
            elif config.separator.model == "openunmix":
                model_sample_rate = separation_model.sample_rate
                sources_order = ["vocals", "drums", "bass", "other"]
                normalise = False
            else:
                raise ValueError(
                    f"Separator model {config.separator.model} not supported."
                )
```

The logic above can be taken outside of the whole song_listener loop as it does not depend on the song or listener.

So the code inside the inner 'if' branch will be much simpler,

```
       if prev_song_name != song_name:
            # Decompose song only once
            prev_song_name = song_name

            sample_rate, mixture_signal = wavfile.read(
                Path(config.path.music_dir)
                / split_directory
                / song_name
                / "mixture.wav"
            )
            mixture_signal = (mixture_signal / 32768.0).astype(np.float32).T
            assert sample_rate == config.nalr.fs

            # Resample mixture signal to model sample rate
            if sample_rate != model_sample_rate:
                resampler = Resample(sample_rate, model_sample_rate)
                mixture_signal = resampler(mixture_signal)

            # Decompose mixture signal into stems
            stems = decompose_signal(
                separation_model,
                mixture_signal,
                sample_rate,
                device,
                sources_order,
                audiogram_left,
                audiogram_right,
                normalise)
```

In v0.4.0 it will be even simpler because the resmapling and float-to-int conversion are handled inside the file reading function.

👍

jonbarker68 reviewed (changes requested):
Nice work. I've got a suggestion for clarifying the 'enhance' function which has become quite long.

groadabike commented:
@jonbarker68 ,
I will submit another PR updating main using the improvements of this version.
I.e., adding the flac and 16 bit things

jonbarker68 commented on recipes/cad1/task1/baseline/enhance.py:
```diff
@@ -372,38 +420,58 @@ def enhance(config: DictConfig) -> None:
         n_samples = processed_stems[list(processed_stems.keys())[0]].shape[0]
         out_left, out_right = np.zeros(n_samples), np.zeros(n_samples)
-        for stem_str, item in processed_stems.items():
+        for stem_str, stem_signal in processed_stems.items():
```
I think it would be clearer if the loop that does the 'remixing' was kept separate from the loop that writes out the stems. i.e.  so that the code has a clearer decomposition into three stages

- demix
- remix
- save stems and remix

It probably also needs to be clear that the stem outputs and the remixed outputs are in different formats. i.e. the stems are saving a scaling constant (so that they can be reconstructed as float) whereas the mix is being clipped to 16bit.

jonbarker68 reviewed (changes requested):
Thanks.  Just a couple of refactoring suggestions for task1 enhance.py which has become rather complicated.

jonbarker68 commented:
Great. That looks much clearer.  There is now just a parameter missing in the call to  decompose_signal in test_enhance_task1.py

groadabike commented:
Thank you,

Yeah, I fixed that error in the test and added the new ones.
Hope is ok for merging


On Thu, 11 May 2023 at 09:47, Jon Barker ***@***.***> wrote:

> Great. That looks much clearer. There is now just a parameter missing in
> the call to decompose_signal in test_enhance_task1.py
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/claritychallenge/clarity/pull/300#issuecomment-1543592575>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AED2G2WERLIG5BXSWLN5T5DXFSRQXANCNFSM6AAAAAAXELDSJY>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>


-- 
Gerardo Roa Dabike
Ingeniero Civil Informatico
